From 4753df51ecf46422ac25d79aab7f79ee4f22b3e9 Mon Sep 17 00:00:00 2001
From: rguenth <rguenth@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Wed, 4 Jun 2014 13:41:09 +0000
Subject: [PATCH 031/134] 2014-06-04  Richard Biener  <rguenther@suse.de>

	PR tree-optimization/61383
	* tree-ssa-ifcombine.c (bb_no_side_effects_p): Make sure
	stmts can't trap.

	* gcc.dg/torture/pr61383-1.c: New testcase.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@211232 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                            |  6 ++++++
 gcc/testsuite/ChangeLog                  |  5 +++++
 gcc/testsuite/gcc.dg/torture/pr61383-1.c | 35 ++++++++++++++++++++++++++++++++
 gcc/tree-ssa-ifcombine.c                 |  4 ++++
 4 files changed, 50 insertions(+)
 create mode 100644 gcc/testsuite/gcc.dg/torture/pr61383-1.c

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 343050c..a6eeb63 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,9 @@
+2014-06-04  Richard Biener  <rguenther@suse.de>
+
+	PR tree-optimization/61383
+	* tree-ssa-ifcombine.c (bb_no_side_effects_p): Make sure
+	stmts can't trap.
+
 2014-06-03  Andrey Belevantsev  <abel@ispras.ru>
 
 	Backport from mainline
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 0bba1d5..b66a7aa 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,8 @@
+2014-06-04  Richard Biener  <rguenther@suse.de>
+
+	PR tree-optimization/61383
+	* gcc.dg/torture/pr61383-1.c: New testcase.
+
 2014-06-03  Andrey Belevantsev  <abel@ispras.ru>
 
 	Backport from mainline
diff --git a/gcc/testsuite/gcc.dg/torture/pr61383-1.c b/gcc/testsuite/gcc.dg/torture/pr61383-1.c
new file mode 100644
index 0000000..d9a0a0b
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/torture/pr61383-1.c
@@ -0,0 +1,35 @@
+/* { dg-do run } */
+
+int a, b = 1, c, d, e, f, g;
+
+int
+fn1 ()
+{
+  int h;
+  for (;;)
+    {
+      g = b;
+      g = g ? 0 : 1 % g;
+      e = a + 1;
+      for (; d < 1; d = e)
+	{
+	  if (f == 0)
+	    h = 0;
+	  else
+	    h = 1 % f;
+	  if (f < 1)
+	    c = 0;
+	  else if (h)
+	    break;
+	}
+      if (b)
+	return 0;
+    }
+}
+
+int
+main ()
+{
+  fn1 ();
+  return 0;
+}
diff --git a/gcc/tree-ssa-ifcombine.c b/gcc/tree-ssa-ifcombine.c
index cc06ca1..ed6ea82 100644
--- a/gcc/tree-ssa-ifcombine.c
+++ b/gcc/tree-ssa-ifcombine.c
@@ -105,7 +105,11 @@ bb_no_side_effects_p (basic_block bb)
     {
       gimple stmt = gsi_stmt (gsi);
 
+      if (is_gimple_debug (stmt))
+	continue;
+
       if (gimple_has_side_effects (stmt)
+	  || gimple_could_trap_p (stmt)
 	  || gimple_vuse (stmt))
 	return false;
     }
-- 
1.8.5.5

